fairseq/criterions/adaptive_loss.py

```diff
@@ -32,13 +32,26 @@ def forward(self, model, sample, reduce=True):
         """
 
         assert hasattr(model.decoder, 'adaptive_softmax') and model.decoder.adaptive_softmax is not None
-        adaptive_softmax = model.decoder.adaptive_softmax
 
         net_output = model(**sample['net_input'])
+
+        loss = self.compute_cross_entropy_loss(model, net_output, sample, reduce)
+
+        sample_size = sample['target'].size(0) if self.args.sentence_avg else sample['ntokens']
+        logging_output = {
+            'loss': utils.item(loss.data) if reduce else loss.data,
+            'ntokens': sample['ntokens'],
+            'sample_size': sample_size,
+        }
+        return loss, sample_size, logging_output
+
+    def compute_cross_entropy_loss(self, model, net_output, sample, reduce=True):
```

Comment:
maybe instead of passing in model, we can pass in adaptive_softmax and target?

Reply:
it will become more complicated if we pass in adaptive_softmax and target, e.g. in the `forward` function of `AdaptiveLoss` and `LabelSmoothedAdaptiveLoss`, we will need to extract the `adaptive_softmax`, extract `target`, which makes the logic more duplicated.

In addition, in this way, the input args for `compute_smooth_loss ` is the same.